csharp/src/Google.Protobuf/FieldMaskTree.cs

```diff
@@ -356,18 +357,51 @@ public void Merge(IMessage source, IMessage destination, FieldMask.MergeOptions
                     }
                     else
                     {
-                        if (sourceField != null
-                            || !options.ReplacePrimitiveFields)
+                        // Primitives can have presence indicated on them, so check that if available
+                        // Otherwise check for default (not null!) (see /docs/field_presence.md#considerations-for-merging)
+                        if ((field.HasPresence && field.Accessor.HasValue(source))
+                            || !IsDefaultValue(field, sourceField))
                         {
                             field.Accessor.SetValue(destination, sourceField);
                         }
-                        else
+                        else if (options.ReplacePrimitiveFields)
```

Comment:
The problem with that is with how FieldMasks are combined, and what happens with parent/child fields.

If you have a field `a.b` that is a non-presence primitive, you're going to get different behavior if you make a fieldmask of `"a"` vs a fieldmask of `"a.b"` because `FieldMask("a")` is going to call `MergeFrom()` which will respect the defaults of `a.b`. Maybe this is intended, but the problem is with the way fieldmasks combine. From the documentation of `FieldMaskTree.AddFieldPath()`:
> ... In a FieldMask, every field path matches the specified field as well as all its sub-fields. For example, a field path "foo.bar" matches field "foo.bar" and also "foo.bar.baz", etc. When adding a field path to the tree, redundant sub-paths will be removed. That is, after adding "foo.bar" to the tree, "foo.bar.baz" will be removed if it exists, which will turn the tree node for "foo.bar" to a leaf node. Likewise, if the field path to add is a sub-path of an existing leaf node, nothing will be changed in the tree.

This behavior seems to strongly indicate that `FieldMask("a")` and `FieldMask("a.b")` should not change the way the field `a.b` is merged.

Reply:
So if you have a message `a.{b,c}`, are `new FieldMaskTree().AddFieldPath("a")` and `new FieldMaskTree().AddFieldPath("a.b").AddFieldPath("a.c")` actually exactly equivalent in terms of default behavior? What about with `options.ReplaceMessageFields = true` and `options.ReplacePrimitiveFields = true`? If not, then is it really correct to have `FieldMaskTree` throw out child paths if they are covered by a parent path?

Reply:
Okay, looking at this again... the more I look at it, the less clear it seems to be.

I propose we take a step back, and try to follow these goals:

- Where class/method documentation is simply unclear, we clarify it; this is a particular concern for the options where the original Java code seems more clear to me than the C# code. (Hopefully we can just use the Java wording, if the semantics are the same.)
- Where the current implementation follows "a reasonable interpretation" of the documentation, we don't change the behavior, but clarify the docs so that less interpretation is required. This avoids breaking changes where a user could reasonably expect the current behavior from the docs (even if it's not the behavior we might want).
- Where there's complete ambiguity about what should happen, we can decide what we want to happen.

Your point about throwing out child paths is a really good and important one, which I *suspect* comes with some limitations which we should mention in the top-level documentation. I suspect that `FieldMaskTree` is only useful as described in specific situations.

Does all of that make sense as an approach? I'll need to spend yet more time reviewing *all* the docs of this - `FieldMaskTree` is one of the bits of Google.Protobuf which I wasn't involved in at all, so I need to get more familiar with it. (And this is *relatively* low-priority work - I'm trying to make time for it when I can, but it's unlikely to jump to the top of my list, I'm afraid.)